### PR TITLE
SRV-26 Select Banners with aduser info

### PR DESCRIPTION
**Pull request by teghnet**

### CODE REVIEW

File: app/Client/DummyAdSelectClient.php [hunk 1]
```diff
@@ -114,4 +110,25 @@ private function getBestBanners(array $zones, array $keywords): array
...
+
+    private function queryBuilder($key, $zone, array $typeDefault): Builder
+    {
+        $queryBuilder = DB::table('network_banners')->join(
+            'network_campaigns',
+            'network_banners.network_campaign_id',
+            '=',
+            'network_campaigns.id'
+        )->select('network_banners.uuid')->whereRaw(
+            '(network_campaigns.targeting_requires LIKE ? '."OR network_campaigns.targeting_requires = '[]')",
```
Comment (PawelPodkalicki) [inline app/Client/DummyAdSelectClient.php]:
```suggestion
            "(network_campaigns.targeting_requires LIKE ? OR network_campaigns.targeting_requires = '[]')",
```

File: app/Client/DummyAdSelectClient.php [hunk 2]
```diff
@@ -114,4 +110,25 @@ private function getBestBanners(array $zones, array $keywords): array
...
+            "%$key%"
+        )->whereRaw(
+            '(network_campaigns.targeting_excludes NOT LIKE ? '."OR network_campaigns.targeting_excludes = '[]')",
```
Comment (PawelPodkalicki) [inline app/Client/DummyAdSelectClient.php]:
```suggestion
            "(network_campaigns.targeting_excludes NOT LIKE ? OR network_campaigns.targeting_excludes = '[]')",
```

File: app/Client/GuzzleAdUserClient.php [hunk 3]
```diff
@@ -25,10 +25,13 @@
...
 use Adshares\Common\Application\Factory\TaxonomyFactory;
 use Adshares\Common\Application\Service\TargetingOptionsSource;
+use Adshares\Supply\Application\Dto\ImpressionContext;
+use Adshares\Supply\Application\Dto\UserContext;
+use Adshares\Supply\Application\Service\UserContextProvider;
 use GuzzleHttp\Client;
 use function GuzzleHttp\json_decode;
 
-final class GuzzleAdUserClient implements TargetingOptionsSource
+final class GuzzleAdUserClient implements TargetingOptionsSource, UserContextProvider
```
Comment (PawelPodkalicki) [inline app/Client/GuzzleAdUserClient.php:34]: GuzzleAdUserClient and DummyAdUserClient have different interface

File: app/Client/JsonRpcAdSelectClient.php [hunk 4]
```diff
@@ -44,15 +51,71 @@ public function __construct(JsonRpc $client)
...
+            } else {
+                yield null;
+            }
+        }
+    }
+
+    private function find(array $bannerIds): Generator
+    {
+        foreach ($bannerIds as $bannerId) {
+            $banner = $bannerId ? NetworkBanner::where('uuid', hex2bin($bannerId))->first() : NetworkBanner::first();
```
Comment (PawelPodkalicki) [inline app/Client/JsonRpcAdSelectClient.php]: If bannerId is not defined. banner should be null/empty.

File: src/Supply/Application/Dto/UserContext.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    /** @var string */
+    private $userId;
+
+    public function __construct(array $keywords, float $humanScore, string $userId)
+    {
+        $this->keywords = $keywords;
+        $this->humanScore = $humanScore;
+        $this->userId = $userId;
+//        $this->failIfInvalid();
```
Comment (PawelPodkalicki) [inline src/Supply/Application/Dto/UserContext.php]: Is this comment intended?

**Review by PawelPodkalicki — CHANGES_REQUESTED** [review]
I don't understand why some files still have different formatting, while we use the same formatting rules.

File: app/Client/GuzzleAdUserClient.php (hunk 3 above)
Comment (teghnet) [reply app/Client/GuzzleAdUserClient.php:34]: Yes. It is Intentional.

File: src/Supply/Application/Dto/UserContext.php (hunk 5 above)
Comment (teghnet) [reply src/Supply/Application/Dto/UserContext.php]: Yes, because the uid has different format for now

**Review by PawelPodkalicki — APPROVED** [review]

**Pull request merged**